catboost/libs/metrics/metric.cpp

```diff
@@ -3166,6 +3166,122 @@ void TAverageGain::GetBestValue(EMetricBestValue* valueType, float*) const {
     *valueType = EMetricBestValue::Max;
 }
 
+/* NormalizedGini */
+
+THolder<TNormalizedGiniMetric> TNormalizedGiniMetric::CreateBinClassMetric(double border) {
+    return MakeHolder<TNormalizedGiniMetric>(border);
+}
+
+THolder<TNormalizedGiniMetric> TNormalizedGiniMetric::CreateMultiClassMetric(int positiveClass) {
+    CB_ENSURE(positiveClass >= 0, "Class id should not be negative");
+
+    auto metric = new TNormalizedGiniMetric();
+    metric->PositiveClass = positiveClass;
+    metric->IsMultiClass = true;
+    return metric;
+}
+
+TMetricHolder TNormalizedGiniMetric::Eval(
+        const TVector<TVector<double>>& approx,
+        TConstArrayRef<float> target,
+        TConstArrayRef<float> weightIn,
+        TConstArrayRef<TQueryInfo> /*queriesInfo*/,
+        int begin,
+        int end,
+        NPar::TLocalExecutor& /* executor */
+) const {
+    Y_ASSERT((approx.size() > 1) == IsMultiClass);
+    const auto& approxVec = approx.ysize() == 1 ? approx.front() : approx[PositiveClass];
+    Y_ASSERT(approxVec.size() == target.size());
+
+    TVector<double> targetCopy(target.begin() + begin, target.begin() + end);
+
+    if (!IsMultiClass) {
+        for (ui32 target : targetCopy) {
+            target = target > Border;
+        }
+    }
+
+    if (approx.ysize() > 1) {
+        for (ui32 target : targetCopy) {
+            target = (target == static_cast<double>(PositiveClass));
+        }
+    }
+
+    TVector<double> approxCopy(approxVec.begin() + begin, approxVec.begin() + end);
+    const auto& weight = UseWeights ? weightIn : TVector<float>{};
+
+    TVector<NMetrics::TSample> samples;
+    if (weight.empty()) {
+        samples = NMetrics::TSample::FromVectors(targetCopy, approxCopy);
+    } else {
+        TVector<double> weightCopy(weight.begin() + begin, weight.begin() + end);
+        samples = NMetrics::TSample::FromVectors(targetCopy, approxCopy, weightCopy);
+    }
+
+    double normalizedGini = 2 * CalcAUC(&samples) - 1;
+
+    TMetricHolder error(2);
+    error.Stats[0] = normalizedGini;
+    error.Stats[1] = 1.0;
+    return error;
+}
+
+TString TNormalizedGiniMetric::GetDescription() const {
+    if (IsMultiClass) {
+        const TMetricParam<int> positiveClass("class", PositiveClass, /*userDefined*/true);
+        return BuildDescription(ELossFunction::NormalizedGini, UseWeights, positiveClass);
+    } else {
```

Comment:
imho, this can be simplified:
```suggestion
     } 
     
     return BuildDescription(ELossFunction::NormalizedGini, UseWeights, MakeBorderParam(Border));
```